### DISCUSSION TITLE
Remove/Refactor/Update "directFormat" in favor of using overwritten contents of .cache file rather than reading from stdout?

### DISCUSSION THREAD

**Discussion by dhananjaipai** [AkiraVoid-Productions/vscode-extension-format-with-cli#14] [Ideas]
I think some formatters only support overwriting the original file - `***/.~format-with-cli.cache` in this case, and not writing the "formatted" output to stdout

For the setting where the formatter writes to stdout, it will also exit with a non-zero exit code. Is there a way to use directFormat setting with the contents of the "updated" .cache file instead of the standard out?

I tried playing around with the command in the case of the `just` formatter so that I can print it to stdout as a workaround  and mask the exit code
```
# just --fmt --unstable --justfile \"{file}\" # overwrites original file

# checks and exits with non-zero exit code
just --fmt --unstable --check --justfile \"{file}\" || true
```
but it ended up not writing the formatted file, but highlighted the things to be change like below (where `-` is lines to remove and `+` is formatted lines to add.
```
 
-
-
-
-
 apply: aws-token merge jinja terraform
 
 @verify BINARY:
-                if ! hash {{ BINARY }} &> /dev/null; then echo "{{ BINARY }} is not installed"; else echo "{{ BINARY }} is installed"; fi
+    if ! hash {{ BINARY }} &> /dev/null; then echo "{{ BINARY }} is not installed"; else echo "{{ BINARY }} is installed"; fi
```

I also tried `"just --fmt --unstable --justfile \"{file}\" && cat \"{file}\""` as a workaround to print it to stdout, but I think the replacement only happens once? got the error as 
```
Command failed: just --fmt --unstable --justfile "***/.~format-with-cli.cache" && cat "{file}"
```

Final workaround was something like 
```
"AkiraVoid.formatWithCLI.command":  "FILE=\"{file}\" && just --fmt --unstable --justfile \"${FILE}\" && cat \"${FILE}\""
```
which worked, but seemed a bit hacky.

Thought I would open an issue for discussion and might help others who are on the same boat with a cli formatter.

_Originally posted by @dhananjaipai in https://github.com/AkiraVoid-Productions/vscode-extension-format-with-cli/issues/10#issuecomment-2241116078_

**Comment by dhananjaipai** [comment]
@AkiraVoid Just saw your comment on the original thread, and it makes sense.
I think the performance impact can be off-putting for people who are trying to format large files.

May be replacing all occurances of `{file}` in the command can be a feature request? Right now it is only replacing the first occurance.

Just so that I do not have to define a shell variable for referencing it in multiple places.

**Reply by AkiraVoid** [reply]
Agree with you, it's a fair suggestion.

**Comment by AkiraVoid** [comment]
Well, if we write the contents to a cache file, format it and then read it out, we actually performed the IO operation for four times (two for extension and 2 for formatter), which is extremely slow when it comes to a large file (I'd say it's even annoying for a small file). Now as you can see, with Direct Format, you don't need to remind yourself save the file first, but it is apparently slower than Direct Format disabled. IO is slow, especially when it comes to Node.js, and I have no confidence on making an IO solution that performs better than standard library…

**Reply by dhananjaipai** [reply]
Yeah, makes sense.

In that case, could I request a solution where we have another flag for performance called with formatStdIN or something?
Technically, for prettier at least, you can [skip the need to write to the file](https://prettier.io/docs/en/cli#--stdin-filepath)
you could run
```bash
echo '{"hello":         "world"}' | prettier --stdin-filepath anything_here.json
```
where `{"hello":         "world"}` is the contents of the file, and it will write the formatted output to stdout
```json
{ "hello": "world" }
```

We are skipping the file write altogether and making the IO faster, but the user needs to have some understanding of the formatter for them to leverage this - like, adding the `--stdin-filepath` for prettier in this case, so perhaps not the default behavior

**Comment by dhananjaipai** [comment]
One implementation suggestion is to change the implementation for performance improvement (without using the writeFileSync to cacheFile may be as below

Replace line :83 from
` const command = getCommand(doc.languageId).replace('{file}', cacheFile); ` to be based on whether cacheFile is enabled or not (using an additional flag).

If cacheFile is not enabled (and you want to do it through stdout), you could do
`` const command = `echo  '${content}' | getCommand(doc.languageId).replaceAll('{file}', doc.fileName)` ``
This will pipe the content into the command through stdin and the command just has to be `prettier --stdin-filepath {file}` we will still need the file since prettier uses the file extension to identify the type I am assuming.

**Comment by AkiraVoid** [comment]
As the comment in #16 , we won't support this as a feature. Alternatively, as we now support using `{file}` placeholder multiple times, users could do `cat {file} | npx prettier --ignore-unknown --stdin-filepath {file}` by themselves.
